// @lowdefy/operators/dist/webParser.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _helpers = require("@lowdefy/helpers");

var _common = _interopRequireDefault(require("./common"));

var _web = _interopRequireDefault(require("./web"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class WebParser {
  constructor(_ref) {
    var {
      context,
      contexts
    } = _ref;
    this.context = context;
    this.contexts = contexts;
    this.parse = this.parse.bind(this);
    this.operations = _objectSpread(_objectSpread({}, _common.default), _web.default);
  }

  parse(_ref2) {
    var {
      args,
      arrayIndices,
      event,
      input,
      location
    } = _ref2;

    if (_helpers.type.isUndefined(input)) {
      return {
        output: input,
        errors: []
      };
    }

    if (event && !_helpers.type.isObject(event)) {
      throw new Error('Operator parser event must be a object.');
    }

    if (args && !_helpers.type.isArray(args)) {
      throw new Error('Operator parser args must be an array.');
    }

    if (location && !_helpers.type.isString(location)) {
      throw new Error('Operator parser location must be a string.');
    }

    var errors = [];

    var reviver = (_, value) => {
      if (_helpers.type.isObject(value) && Object.keys(value).length === 1) {
        var key = Object.keys(value)[0];
        var [op, methodName] = key.split('.');

        try {
          if (!_helpers.type.isUndefined(this.operations[op])) {
            var res = this.operations[op]({
              eventLog: this.context.eventLog,
              args,
              arrayIndices,
              context: this.context,
              contexts: this.contexts,
              env: 'web',
              event,
              input: this.context.input,
              location: location ? (0, _helpers.applyArrayIndices)(arrayIndices, location) : null,
              lowdefyGlobal: this.context.lowdefyGlobal,
              menus: this.context.menus,
              methodName,
              operations: this.operations,
              params: value[key],
              requests: this.context.requests,
              state: this.context.state,
              urlQuery: this.context.urlQuery,
              parser: this
            });
            return res;
          }
        } catch (e) {
          errors.push(e);
          console.error(e);
          return null;
        }
      }

      return value;
    };

    return {
      output: _helpers.serializer.copy(input, {
        reviver
      }),
      errors
    };
  }

}

var _default = WebParser;
exports.default = _default;
