// @lowdefy/nunjucks/dist/nunjucks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.nunjucksFunction = exports.validNunjucksString = exports.nunjucksString = exports.nunjucksEnv = void 0;

var _nunjucks = _interopRequireDefault(require("nunjucks"));

var _nunjucksDateFilter = _interopRequireDefault(require("nunjucks-date-filter"));

var _helpers = require("@lowdefy/helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
  Copyright 2020-2021 Lowdefy, Inc

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/
// dateFilter.setDefaultFormat('YYYY-MM-DD');
var nunjucksEnv = new _nunjucks.default.Environment();
exports.nunjucksEnv = nunjucksEnv;
nunjucksEnv.addFilter('date', _nunjucksDateFilter.default);
var nunjucksTemplates = {}; // slow

var nunjucksString = (templateString, value) => {
  if (_helpers.type.isPrimitive(value)) {
    return nunjucksEnv.renderString(templateString, {
      value
    });
  }

  return nunjucksEnv.renderString(templateString, value);
};

exports.nunjucksString = nunjucksString;

var validNunjucksString = function validNunjucksString(templateString) {
  var returnError = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

  try {
    nunjucksString(templateString, {});
    return true;
  } catch (e) {
    if (returnError) {
      return {
        name: e.name,
        message: e.message
      };
    }

    return false;
  }
}; // fast
// test with memoization
// this method compiles a nunjucks string only if the client has not compiled the same string before.


exports.validNunjucksString = validNunjucksString;

var nunjucksFunction = templateString => {
  // template was already compiled
  if (_helpers.type.isFunction(nunjucksTemplates[templateString])) {
    return nunjucksTemplates[templateString];
  }

  if (_helpers.type.isString(templateString)) {
    var template = _nunjucks.default.compile(templateString, nunjucksEnv); // execute once to throw catch template errors


    template.render({});

    nunjucksTemplates[templateString] = value => {
      if (_helpers.type.isPrimitive(value)) {
        return template.render({
          value
        });
      }

      return template.render(value);
    };
  } else {
    // for non string types like booleans or objects
    nunjucksTemplates[templateString] = () => templateString;
  }

  return nunjucksTemplates[templateString];
};

exports.nunjucksFunction = nunjucksFunction;
