// @lowdefy/operators/dist/common/date.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _runClass = _interopRequireDefault(require("../runClass"));

var _helpers = require("@lowdefy/helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
  Copyright 2020-2021 Lowdefy, Inc

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/
function date(input) {
  var result = new Date(input);

  if (!_helpers.type.isDate(result)) {
    throw new Error("".concat(input, " could not resolve as a valid javascript date."));
  }

  return result;
}

function now() {
  return new Date();
}

var functions = {
  __default: date,
  now
};
var meta = {
  __default: {
    singleArg: true,
    validTypes: ['number', 'string']
  },
  now: {
    noArgs: true
  }
};

function _date(_ref) {
  var {
    params,
    location,
    methodName
  } = _ref;
  return (0, _runClass.default)({
    functions,
    location,
    meta,
    methodName: methodName,
    operator: '_date',
    params,
    defaultFunction: '__default'
  });
}

var _default = _date;
exports.default = _default;
