// mingo/operators/_predicates.js
"use strict";
/**
 * Predicates used for Query and Expression operators.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.$type = exports.$elemMatch = exports.$size = exports.$all = exports.$exists = exports.$regex = exports.$mod = exports.$gte = exports.$gt = exports.$lte = exports.$lt = exports.$nin = exports.$in = exports.$ne = exports.$eq = exports.createExpressionOperator = exports.createQueryOperator = void 0;
var core_1 = require("../core");
var query_1 = require("../query");
var util_1 = require("../util");
/**
 * Returns a query operator created from the predicate
 *
 * @param pred Predicate function
 */
function createQueryOperator(pred) {
    return function (selector, value, options) {
        var opts = { unwrapArray: true };
        return function (obj) {
            // value of field must be fully resolved.
            var lhs = util_1.resolve(obj, selector, opts);
            return pred(lhs, value, options);
        };
    };
}
exports.createQueryOperator = createQueryOperator;
/**
 * Returns an expression operator created from the predicate
 *
 * @param f Predicate function
 */
function createExpressionOperator(f) {
    return function (obj, expr, options) {
        var args = core_1.computeValue(obj, expr, null, options);
        return f.apply(void 0, args);
    };
}
exports.createExpressionOperator = createExpressionOperator;
/**
 * Checks that two values are equal.
 *
 * @param a         The lhs operand as resolved from the object by the given selector
 * @param b         The rhs operand provided by the user
 * @returns {*}
 */
function $eq(a, b, options) {
    // start with simple equality check
    if (util_1.isEqual(a, b))
        return true;
    // https://docs.mongodb.com/manual/tutorial/query-for-null-fields/
    if (util_1.isNil(a) && util_1.isNil(b))
        return true;
    // check
    if (a instanceof Array) {
        var eq = util_1.isEqual.bind(null, b);
        return a.some(eq) || util_1.flatten(a, 1).some(eq);
    }
    return false;
}
exports.$eq = $eq;
/**
 * Matches all values that are not equal to the value specified in the query.
 *
 * @param a
 * @param b
 * @returns {boolean}
 */
function $ne(a, b, options) {
    return !$eq(a, b, options);
}
exports.$ne = $ne;
/**
 * Matches any of the values that exist in an array specified in the query.
 *
 * @param a
 * @param b
 * @returns {*}
 */
function $in(a, b, options) {
    // queries for null should be able to find undefined fields
    if (util_1.isNil(a))
        return b.some(util_1.isNull);
    return util_1.intersection(util_1.ensureArray(a), b, options === null || options === void 0 ? void 0 : options.hashFunction).length > 0;
}
exports.$in = $in;
/**
 * Matches values that do not exist in an array specified to the query.
 *
 * @param a
 * @param b
 * @returns {*|boolean}
 */
function $nin(a, b, options) {
    return !$in(a, b, options);
}
exports.$nin = $nin;
/**
 * Matches values that are less than the value specified in the query.
 *
 * @param a
 * @param b
 * @returns {boolean}
 */
function $lt(a, b, options) {
    return compare(a, b, function (x, y) { return x < y; });
}
exports.$lt = $lt;
/**
 * Matches values that are less than or equal to the value specified in the query.
 *
 * @param a
 * @param b
 * @returns {boolean}
 */
function $lte(a, b, options) {
    return compare(a, b, function (x, y) { return x <= y; });
}
exports.$lte = $lte;
/**
 * Matches values that are greater than the value specified in the query.
 *
 * @param a
 * @param b
 * @returns {boolean}
 */
function $gt(a, b, options) {
    return compare(a, b, function (x, y) { return x > y; });
}
exports.$gt = $gt;
/**
 * Matches values that are greater than or equal to the value specified in the query.
 *
 * @param a
 * @param b
 * @returns {boolean}
 */
function $gte(a, b, options) {
    return compare(a, b, function (x, y) { return x >= y; });
}
exports.$gte = $gte;
/**
 * Performs a modulo operation on the value of a field and selects documents with a specified result.
 *
 * @param a
 * @param b
 * @returns {boolean}
 */
function $mod(a, b, options) {
    return util_1.ensureArray(a).some(function (x) { return b.length === 2 && x % b[0] === b[1]; });
}
exports.$mod = $mod;
/**
 * Selects documents where values match a specified regular expression.
 *
 * @param a
 * @param b
 * @returns {boolean}
 */
function $regex(a, b, options) {
    var lhs = util_1.ensureArray(a);
    var match = function (x) { return util_1.isString(x) && !!b.exec(x); };
    return lhs.some(match) || util_1.flatten(lhs, 1).some(match);
}
exports.$regex = $regex;
/**
 * Matches documents that have the specified field.
 *
 * @param a
 * @param b
 * @returns {boolean}
 */
function $exists(a, b, options) {
    return (((b === false || b === 0) && a === undefined) ||
        ((b === true || b === 1) && a !== undefined));
}
exports.$exists = $exists;
/**
 * Matches arrays that contain all elements specified in the query.
 *
 * @param a
 * @param b
 * @returns boolean
 */
function $all(a, b, options) {
    var matched = false;
    if (util_1.isArray(a) && util_1.isArray(b)) {
        for (var i = 0, len = b.length; i < len; i++) {
            if (util_1.isObject(b[i]) && util_1.inArray(Object.keys(b[i]), "$elemMatch")) {
                matched = matched || $elemMatch(a, b[i]["$elemMatch"], options);
            }
            else {
                // order of arguments matter
                return util_1.intersection(b, a, options === null || options === void 0 ? void 0 : options.hashFunction).length === len;
            }
        }
    }
    return matched;
}
exports.$all = $all;
/**
 * Selects documents if the array field is a specified size.
 *
 * @param a
 * @param b
 * @returns {*|boolean}
 */
function $size(a, b, options) {
    return a.length === b;
}
exports.$size = $size;
function isNonBooleanOperator(name) {
    return util_1.isOperator(name) && ["$and", "$or", "$nor"].indexOf(name) === -1;
}
/**
 * Selects documents if element in the array field matches all the specified $elemMatch condition.
 *
 * @param a {Array} element to match against
 * @param b {Object} subquery
 */
function $elemMatch(a, b, options) {
    // should return false for non-matching input
    if (util_1.isArray(a) && !util_1.isEmpty(a)) {
        var format = function (x) { return x; };
        var criteria = b;
        // If we find a boolean operator in the subquery, we fake a field to point to it. This is an
        // attempt to ensure that it is a valid criteria. We cannot make this substitution for operators
        // like $and/$or/$nor; as otherwise, this faking will break our query.
        if (Object.keys(b).every(isNonBooleanOperator)) {
            criteria = { temp: b };
            format = function (x) { return ({ temp: x }); };
        }
        var query = new query_1.Query(criteria, options);
        for (var i = 0, len = a.length; i < len; i++) {
            if (query.test(format(a[i]))) {
                return true;
            }
        }
    }
    return false;
}
exports.$elemMatch = $elemMatch;
/**
 * Selects documents if a field is of the specified type.
 *
 * @param a
 * @param b
 * @returns {boolean}
 */
function $type(a, b, options) {
    switch (b) {
        case 1:
        case 19:
        case util_1.BsonType.DOUBLE:
        case util_1.BsonType.DECIMAL:
            return util_1.isNumber(a);
        case 2:
        case util_1.JsType.STRING:
            return util_1.isString(a);
        case 3:
        case util_1.JsType.OBJECT:
            return util_1.isObject(a);
        case 4:
        case util_1.JsType.ARRAY:
            return util_1.isArray(a);
        case 6:
        case util_1.JsType.UNDEFINED:
            return util_1.isNil(a);
        case 8:
        case util_1.JsType.BOOLEAN:
        case util_1.BsonType.BOOL:
            return util_1.isBoolean(a);
        case 9:
        case util_1.JsType.DATE:
            return util_1.isDate(a);
        case 10:
        case util_1.JsType.NULL:
            return util_1.isNull(a);
        case 11:
        case util_1.JsType.REGEXP:
        case util_1.BsonType.REGEX:
            return util_1.isRegExp(a);
        case 16:
        case util_1.BsonType.INT:
            return (util_1.isNumber(a) &&
                a >= util_1.MIN_INT &&
                a <= util_1.MAX_INT &&
                a.toString().indexOf(".") === -1);
        case 18:
        case util_1.BsonType.LONG:
            return (util_1.isNumber(a) &&
                a >= util_1.MIN_LONG &&
                a <= util_1.MAX_LONG &&
                a.toString().indexOf(".") === -1);
        default:
            return false;
    }
}
exports.$type = $type;
function compare(a, b, f) {
    return util_1.ensureArray(a).some(function (x) { return util_1.getType(x) === util_1.getType(b) && f(x, b); });
}
